src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs

```diff
@@ -0,0 +1,135 @@
+// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using Elastic.CommonSchema;
+using Elasticsearch.Net;
+
+namespace Elasticsearch.Extensions.Logging
+{
+	internal class ElasticsearchDataShipper : IDisposable
+	{
+		private const int MaxQueuedMessages = 1024;
+
+		private readonly BlockingCollection<LogEvent> _messageQueue =
+			new BlockingCollection<LogEvent>(MaxQueuedMessages);
+
+		private readonly Thread _outputThread;
+
+		public ElasticsearchDataShipper()
+		{
+			_outputThread = new Thread(ProcessLogQueue) { IsBackground = true, Name = $"{nameof(ElasticsearchLogger)}.{nameof(ProcessLogQueue)}" };
+			_outputThread.Start();
+		}
+
+		private IElasticLowLevelClient _lowLevelClient = default!;
+
+		private ElasticsearchLoggerOptions _options = default!;
+
+		internal ElasticsearchLoggerOptions Options
+		{
+			get => _options;
+			set
+			{
+				_options = value;
+				UpdateClient();
+			}
+		}
+
+		public void Dispose()
+		{
+			_messageQueue?.CompleteAdding();
+			try
+			{
+				_outputThread.Join(1500); // with timeout in case writer is locked
+			}
+			catch (ThreadStateException) { }
```

Comment:
is there an _internal_ logger to which this exception can be written?

Reply:
I looked into fallback loggers in the Mic.Ext.Logging world but there is no such concept. The best we can do is expose callbacks to allow to user to log manually to fallback locations in their application configuration.

This is a carry over of #85 (introduced there not here) that is rewritten in #98.